test/hepmc/edm4hep_testhepmc.cc

```diff
@@ -0,0 +1,130 @@
+      //         \_v3_/       \                           #
+      //         /    \        p8                         #
+      //    v2__p4     \                                  #
+      //   /            p6                                #
+      // p2                                               #
+      //                                                  #
+
+      // First create the event container, with Signal Process 20, event number 1
+      //
+      std::unique_ptr<GenEvent> evt = std::make_unique<GenEvent>( 20, 1 );
+      // define the units
+      evt->use_units(HepMC::Units::GEV, HepMC::Units::MM);
+      //
+      // create vertex 1 and vertex 2, together with their inparticles
+      GenVertex* v1 = new GenVertex();
+      evt->add_vertex( v1 );
+      v1->add_particle_in( new GenParticle( FourVector(0,0,7000,7000),
+                 2212, 3 ) );
+      GenVertex* v2 = new GenVertex();
+      evt->add_vertex( v2 );
+      v2->add_particle_in( new GenParticle( FourVector(0,0,-7000,7000),
+                 2212, 3 ) );
+      //
+      // create the outgoing particles of v1 and v2
+      GenParticle* p3 = 
+    new GenParticle( FourVector(.750,-1.569,32.191,32.238), 1, 3 );
+      v1->add_particle_out( p3 );
+      GenParticle* p4 = 
+    new GenParticle( FourVector(-3.047,-19.,-54.629,57.920), -2, 3 );
+      v2->add_particle_out( p4 );
+      //
+      // create v3
+      GenVertex* v3 = new GenVertex();
+      evt->add_vertex( v3 );
+      v3->add_particle_in( p3 );
+      v3->add_particle_in( p4 );
+      v3->add_particle_out( 
+    new GenParticle( FourVector(-3.813,0.113,-1.833,4.233 ), 22, 1 )
+    );
+      GenParticle* p5 = 
+    new GenParticle( FourVector(1.517,-20.68,-20.605,85.925), -24,3);
+      v3->add_particle_out( p5 );
+      //
+      // create v4
+      GenVertex* v4 = new GenVertex(FourVector(0.12,-0.3,0.05,0.004));
+      evt->add_vertex( v4 );
+      v4->add_particle_in( p5 );
+      v4->add_particle_out( 
+    new GenParticle( FourVector(-2.445,28.816,6.082,29.552), 1,1 )
+    );
+      v4->add_particle_out( 
+    new GenParticle( FourVector(3.962,-49.498,-26.687,56.373), -2,1 )
+    );
+      //    
+      // tell the event which vertex is the signal process vertex
+      evt->set_signal_process_vertex( v3 );
+      // the event is complete, we now print it out to the screen
+      evt->print();
+
+      // now clean-up by deleteing all objects from memory
+      //
+      // deleting the event deletes all contained vertices, and all particles
+      // contained in those vertices
+
+
+
+  //Part2: Convert the particles and write to file
+    auto store = podio::EventStore();
+    auto writer = podio::ROOTWriter("edm4hep_testhepmc.root", &store);
+    auto& particles       = store.create<edm4hep::MCParticleCollection>("testparticles");
+    writer.registerForWrite("testparticles");
+    for (auto particle_i = evt->particles_begin(); particle_i != evt->particles_end(); ++particle_i) {
```

Comment:
is 
```suggestion
    for (auto& particle: particles) {
```
Also possible?

Reply:
Hi Andre, I was looking for that as well, but unfortunately HepMC doesn't use a standard container but some kind of self-made iterator, as you mention: http://lcgapp.cern.ch/project/simu/HepMC/205/html/classHepMC_1_1GenEvent.html#37646bf8f35efdcad9e9cb99d4afcf52
If you find  a nicer way let me know!

Reply:
We have the same issue in PODIO: our iterators for relations and vector members are also not standard c++,e.g.:
```
  void addParent(edm4hep::ConstMCParticle);
  unsigned int parents_size() const;
  edm4hep::ConstMCParticle getParents(unsigned int) const;
  std::vector<edm4hep::ConstMCParticle>::const_iterator parents_begin() const;
  std::vector<edm4hep::ConstMCParticle>::const_iterator parents_end() const;
```
In plcio  have introduced a small helper vector_view that can be used in range based for loops: https://stash.desy.de/projects/IL/repos/plcio/browse/source/include/vector_view.h
with this and 
```
Extracode:
vector_view<ConstMCParticle> getParents() const { return vector_view<ConstMCParticle>( parents_begin(), parents_end()) ; } \n
```
one can write range based for loops:
```
for(auto d : mcp.getParents() ){ ...} 
```

If you think that would be useful, we could add it to PODIO as a feature...

Reply:
I think that would be very nice to have